app/code/Magento/Theme/Controller/Result/AsyncCssPlugin.php

```diff
@@ -32,48 +35,94 @@ public function __construct(ScopeConfigInterface $scopeConfig)
     }
 
     /**
-     * Load CSS asynchronously if it is enabled in configuration.
+     * Extracts styles to head after critical css if critical path feature is enabled.
      *
-     * @param Http $subject
-     * @return void
+     * @param Layout $subject
+     * @param Layout $result
+     * @param HttpResponseInterface|ResponseInterface $httpResponse
+     * @return Layout (That should be void, actually)
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
      */
-    public function beforeSendResponse(Http $subject): void
+    public function afterRenderResult(Layout $subject, Layout $result, ResponseInterface $httpResponse)
     {
-        $content = $subject->getContent();
+        if (!$this->isCssCriticalEnabled()) {
+            return $result;
+        }
 
-        if (\is_string($content) && strpos($content, '</body') !== false && $this->scopeConfig->isSetFlag(
-            self::XML_PATH_USE_CSS_CRITICAL_PATH,
-            ScopeInterface::SCOPE_STORE
-        )) {
-            $cssMatches = [];
-            // add link rel preload to style sheets
-            $content = preg_replace_callback(
-                '@<link\b.*?rel=("|\')stylesheet\1.*?/>@',
-                function ($matches) use (&$cssMatches) {
-                    $cssMatches[] = $matches[0];
-                    preg_match('@href=("|\')(.*?)\1@', $matches[0], $hrefAttribute);
-                    $href = $hrefAttribute[2];
-                    if (preg_match('@media=("|\')(.*?)\1@', $matches[0], $mediaAttribute)) {
-                        $media = $mediaAttribute[2];
-                    }
-                    $media = $media ?? 'all';
-                    $loadCssAsync = sprintf(
-                        '<link rel="preload" as="style" media="%s"' .
-                         ' onload="this.onload=null;this.rel=\'stylesheet\'"' .
-                        ' href="%s" />',
-                        $media,
-                        $href
-                    );
-
-                    return $loadCssAsync;
-                },
-                $content
-            );
+        $content = (string)$httpResponse->getContent();
+        $headCloseTag = '</head>';
+
+        $headEndTagFound = strrpos($content, $headCloseTag) !== false;
```

Comment:
From my experience `<body>` tag usually is much bigger than `<head>`, so I suggest to use `strpos` function instead
```suggestion
        $headEndTagFound = strpos($content, $headCloseTag) !== false;
```